### PR TITLE
CAP-212 object store access

### PR DESCRIPTION
**Pull request by crbunney**
* Created  `DataSet` and `MetaData` to encapsulate datasets and metadata respectively
 * `MetaData` class only stores the dataset's name for now, but intention is to add all the required fields
* Created `serialisation` package with:
  * `Serialiser` abstract base class
  * Abstract `MetaDataSerialiser` for classes that serialise `MetaData` objects to/from different formats
  * `MetaDataYamlSerialiser` which serialises `MetaData` classes specifically to/from YAML format
* Created `S3FSDataSetStore` class to encapsulate logic for interacting with files stored in object store
* Created jupyter notebook to demonstrate how to use the `S3FSDataSetStore` class

Then, subsequently, separated the Metadata access from the DataSetStore to make it possible to access metadata without downloading all the data.

### CODE REVIEW

File: clean_air/data/data_store_access.ipynb [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.8.12"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 1
+}
```
Comment (corinnebosley) [inline clean_air/data/data_store_access.ipynb:354]: Beautiful.  I will have to look at the actual notebooks at some point as well as the source code, but I love all the information you have put in here and the way it is organized.  Looks like it makes it easy for beginners and experts alike.

File: clean_air/data/storage.py [hunk 2]
```diff
@@ -198,3 +181,211 @@ def put(self, item: AURNSite) -> None:
...
+
+    def get(self, dataset_id: str) -> Metadata:
+        metadata_key = self._to_s3_key(dataset_id)
+        cached_metadata_file = self._to_cached_file_path(dataset_id)
+
+        try:
+            self._fs.get(metadata_key, str(cached_metadata_file))
+        except PermissionError:
+            msg = f"PermissionError when accessing {metadata_key}."
+            msg += " Object may not exist, or you may have incorrect/misconfigured credentials"
```
Comment (corinnebosley) [inline clean_air/data/storage.py:302]: I didn't know you could do this `+=` things with error messages, I like it.  And also your very clear error messages, I like those too.

File: clean_air/data/storage.py [hunk 3]
```diff
@@ -198,3 +181,211 @@ def put(self, item: AURNSite) -> None:
...
+    s3_args = {"endpoint_url": endpoint_url}
+    if anon:
+        s3_args["config"] = Config(signature_version=UNSIGNED)
+
+    s3 = boto3.resource("s3", **s3_args)
+    bucket = s3.Bucket(bucket_name)
+    data_file_obj = bucket.Object(data_file_path)
+    try:
+        s3.meta.client.head_bucket(Bucket=bucket.name)
+        # data_file_obj.load()  # This fails, not clear why, permissions? It uses Head Object API call under the hood
```
Comment (corinnebosley) [inline clean_air/data/storage.py:350]: Is this a comment from a previous section waiting to be removed, or is it here as a reminder maybe?

File: clean_air/models.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    files: List[Path] = None
+    metadata: Metadata = None
+
+    @property
+    def name(self):
+        return self.metadata.dataset_name if self.metadata else ""
+
+    @property
+    def id(self):
+        return self.metadata.id if self.metadata else None
```
Comment (corinnebosley) [inline clean_air/models.py:59]: I don't think I really understand this one; I feel a bit unsure about the metadata class here and others as I feel this is setting values which may be different on input, but I also think lots more is happening elsewhere that I don't understand, so I think we should go over this file when we can.

File: clean_air/serialisation.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                "contacts": [],  # TODO
+            }
+        )
+
+    def deserialise(self, serialised_obj: str) -> Metadata:
+        obj_dict = yaml.safe_load(serialised_obj)
+        obj_dict["extent"] = shapely.wkt.loads(obj_dict["extent"])
+        obj_dict["crs"] = pyproj.CRS.from_wkt(obj_dict["crs"])
+        obj_dict["data_type"] = DataType(obj_dict["data_type"])
+        return Metadata(**obj_dict)
```
Comment (corinnebosley) [inline clean_air/serialisation.py:53]: Again, don't really understand what's going on here, so it'd be nice if you could talk me thorugh this file.

File: tests/unit/data/test_storage.py [hunk 6]
```diff
@@ -278,3 +285,204 @@ def test_all_data_cached(self):
...
+        # This is a little bit circular, but makes it easier to set up tests for download operations
+        # later, since the paths will match
+        data_dir = Path(self.dataset_store._cache_dir) / Path(test_metadata.id)
+        data_files = [data_dir / p for p in [Path("test-file.csv")]]
+        self.test_dataset = DataSet(data_files, test_metadata)
+
+    def test_available_datasets(self):
+        """
+        WHEN available_datasets is called
+        THEN a list of datasets that exist in the storage bucket is returned
```
Comment (corinnebosley) [inline tests/unit/data/test_storage.py:313]: This is a really nice way to write descriptions for tests.  I might adopt this at some point.

File: tests/unit/data/test_storage.py [hunk 7]
```diff
@@ -278,3 +285,204 @@ def test_all_data_cached(self):
...
+        self.mock_fs.ls.return_value = [f"{self.mock_storage_bucket_name}/{ds}" for ds in expected_datasets]
+
+        actual_datasets = self.dataset_store.available_datasets()
+        self.assertEqual(expected_datasets, actual_datasets)
+
+    def test_get(self):
+        """
+        GIVEN a valid ID for a dataset that exists on the object store
+        WHEN get is called
+        THEN that dataset is returned
```
Comment (corinnebosley) [inline tests/unit/data/test_storage.py:325]: It also ensures that only one test (assert) is included per test, which is something Barnaby and I have debated quite a lot.  We decided in the end that several asserts per test is preferable, but I never thought it was really. I just wanted to stop arguing about it.

File: tests/unit/data/test_storage.py [hunk 8]
```diff
@@ -278,3 +285,204 @@ def test_all_data_cached(self):
...
+
+        actual = self.dataset_store.get(self.test_dataset.id)
+
+        self.assertEqual(self.test_dataset, actual)
+        self.mock_metadata_store.get.assert_called_with(self.test_dataset.id)
+        self.mock_fs.get.assert_called_with(
+            f"{self.mock_storage_bucket_name}/{self.test_dataset.id}/",
+            f"{self.dataset_store._cache_dir}/{self.test_dataset.id}",
+            recursive=True
+        )
```
Comment (corinnebosley) [inline tests/unit/data/test_storage.py:346]: However, I see you have multiple asserts in here.  The debate continues...

File: tests/unit/test_models.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        test_contacts = [ContactDetails("Mr", "John", "West", "johnwest@example.com")]
+
+        m = Metadata(test_dataset_name, test_extent, test_crs, test_description, test_data_type, test_contacts)
+
+        self.assertEqual(test_dataset_name, m.dataset_name)
+        self.assertEqual(test_extent, m.extent)
+        self.assertEqual(test_data_type, m.data_type)
+        self.assertEqual(test_description, m.description)
+        self.assertEqual(test_contacts, m.contacts)
+        self.assertEqual(test_crs, m.crs)
```
Comment (corinnebosley) [inline tests/unit/test_models.py:37]: This is a really interesting structure, I like it.

**Review by corinnebosley — APPROVED** [review]
Well it looks good to me, I have obviously left comments indicating the bits I don't understand, but I can't see any actual problems in here.  I approve.

File: clean_air/models.py (hunk 4 above)
Comment (crbunney) [reply clean_air/models.py:59]: it's basically a convenience method that's equivalent to calling `my_dataset.metadata.id` (but returns `None` if the dataset has no metadata associated with it for some reason).

It seemed a bit odd not to have an `.id` property on dataclass, but I wanted to ensure consistency between this property and the metadata instance

File: clean_air/data/storage.py (hunk 3 above)
Comment (crbunney) [reply clean_air/data/storage.py:350]: sorta.
It's the answer to the question "why are we only checking access to the Bucket?"

File: .gitignore [hunk 10]
```diff
@@ -1,13 +1,114 @@
...
+# mypy
+.mypy_cache/
+.dmypy.json
+dmypy.json
+
 
-# Protect against temporary/user-specific artefacts
-.idea
-venv
+# End of https://www.toptal.com/developers/gitignore/api/python,jupyternotebooks
```
[the remark this answers is not in the document]
Comment (crbunney) [reply .gitignore:114]: Wish I could take credit for it! :)

File: clean_air/data/storage.py (hunk 2 above)
Comment (crbunney) [reply clean_air/data/storage.py:302]: yea, python's operator overloading allows type-appropriate behaviour to be implemented. Similar behaviour is implemented for lists as well, for example.

You can actually add bespoke behaviour for any of python's operators by implementing "special" methods on your classes, such as `__add__` to implement the `+` operator.

I believe pretty much any operator can be overloaded. https://www.programiz.com/python-programming/operator-overloading is a reasonable intro, and the definitive reference is https://docs.python.org/3/reference/datamodel.html#special-method-names

I think it's most common to implement "to string", equality, and comparison methods though using things like `__str__`, `__eq__`, `__lt__`, `__lte__` for example

File: tests/unit/data/test_storage.py (hunk 6 above)
Comment (crbunney) [reply tests/unit/data/test_storage.py:313]: I shamelessly stole it from the Behaviour Driven Development style of describing test cases :)

File: tests/unit/data/test_storage.py (hunk 8 above)
Comment (crbunney) [reply tests/unit/data/test_storage.py:346]: I aim to test a single "thing" per test case, but sometimes validating that "thing" involves multiple asserts.

Here, for example, the "thing" is getting a dataset involves making the expected calls to the metadata store and object store.

Perhaps multiple vs single asserts is a proxy discussion for what is the appropriate scope for a test case? (i.e. what constitutes a "thing")

Comment (corinnebosley) [reply tests/unit/data/test_storage.py:346]: I like the approach of testing one "thing" per test, including as many asserts as is necessary to test that one thing.  But yes, we shall leave this debate for another day.

**Comment by crbunney** [comment]
I've found an issue with the packaging: the new modules I added don't seem to be included in the pip installable package artifact (so they aren't available unless you are checking out the source and manually adding it to the `PYTHONPATH`)

I've set this back to draft whilst I investigate and work on whatever fix is needed

**Comment by crbunney** [comment]
OK, that's fixed now.
The issue was the way the packaging was set up, we had to add each new module manually to an `__init__.py` file for it to be included in the packaged library.

I've done two things in response:
* changed the packaging setup so that it includes ever module in the `clean_air` source code directories by default
* Introduced `tox` and reorganised the code a little so the unittests will be run against the packaged and installed code, not the unpackaged source code. This means that we'll detect any packaging related issues like this in the future

It does mean I've made some structural changes that will affect @corinnebosley and @lucascatherine when they come to merge any branches started before this is merged, but I also know the git-fu to resolve that

File: environment.yml [hunk 11]
```diff
@@ -7,21 +7,22 @@ dependencies:
...
-  - folium
 
   # Dev
   - pytest
   - flake8
   - boto3-stubs-essential
   - pip:
+    - tox<4.0.0 # The conda package didn't install correctly, but using the pip package did
     - moto>=2.2.16
+    - -e .  # Install the source code in development mode. The argument is the path to the folder containing teh setup.cfg
```
Comment (DPeterK) [inline environment.yml:28]: Comment typo: 'teh'. Super minor, but given that I spotted it...

File: src/clean_air/data/data_store_access.ipynb [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+    "Access to stored datasets and metadata are via the `S3FSDataSetStore` and `S3FSMetdataStore` classes respectively.\n",
+    "(`s3fs` is the underlying library used to access the object store)\n",
+    "\n",
+    "Both data access classes provide a similar interface, working with their respective data objects."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "pycharm": {
```
Comment (DPeterK) [inline src/clean_air/data/data_store_access.ipynb:23]: I'm guessing you created this notebook using PyCharm?!

File: src/clean_air/data/data_store_access.ipynb [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+    "## Creating a DataSetStore or MetadataStore\n",
+    "### Helper Functions\n",
+    "Helper functions are available that create instances of both `S3FSDataSetStore` and `S3FSMetadataStore`:\n",
+    "* `create_dataset_store`\n",
+    "* `create_metadata_store`"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
```
Comment (DPeterK) [inline src/clean_air/data/data_store_access.ipynb:65]: Jupyter notebooks are horrible to store in version control. The best practice I've seen in other projects is to always store unexecuted notebooks (that is, where all code cell output has been cleared). That way the only changes that should be in modified notebooks are actual changes to the notebook, not inline metadata like this. Of course different might be being done in this project - this is just a thought!

File: tests/unit/data/test_storage.py (hunk 8 above)
Comment (DPeterK) [reply tests/unit/data/test_storage.py:346]: I'm with @crbunney here - in my test I want to confirm a single "thing" (behaviour, perhaps?) but sometimes that behaviour is complex enough to require multiple assertions. You could argue then that the unit you're testing is too large with too complex behaviour, but to follow that path leads to near infinite tiny functions that are impossible to keep track of, which is more distasteful than multiple assertions per test IMO. 

The main reason for pushing for one assertion per test is that a failing assertion early in a test can mask another, potentially more severe failure later on in the test code...

**Review by DPeterK — APPROVED** [review]
Light touch review from me - but all looks good here @crbunney.

File: src/clean_air/data/data_store_access.ipynb (hunk 12 above)
Comment (crbunney) [reply src/clean_air/data/data_store_access.ipynb:23]: How did you guess? 😄

File: src/clean_air/data/data_store_access.ipynb (hunk 13 above)
Comment (crbunney) [reply src/clean_air/data/data_store_access.ipynb:65]: In this case, the notebook is acting as documentation, so having the result is desirable. Although I would prefer it if the execution count was always `1`

If we do want to strip out this stuff though, this looks like a handy way of automating it: https://zhauniarovich.com/post/2020/2020-10-clearing-jupyter-output-p3/

File: environment.yml (hunk 11 above)
Comment (crbunney) [reply environment.yml]: no problem, fixed now

**Review by corinnebosley — APPROVED** [review]
I'm still a little unclear on whether we continue writing tests using pytest or whether we rewrite them all with tox, but either way this looks good to me.  Nice work.

File: .github/workflows/tests.yaml [hunk 14]
```diff
@@ -68,5 +69,5 @@ jobs:
           # Machine-readable (for creating an exact copy of this environment)
           conda list --explicit
 
-      - name: Run pytest
-        run: pytest --sampledir .sample-data
+      - name: Run unit tests using tox
```
[the remark this answers is not in the document]
Comment (crbunney) [reply .github/workflows/tests.yaml:72]: Ahh good question!
Pytest is still running the tests. What tox is doing is orchestrating the process of building the source code into the pip installable package and installing it into an isolated virtualenv with its dependencies. Once installed it runs pytest and collects the result. Line 8 of the `tox.ini` file is where tox is configured to invoke pytest

[the remark this answers is not in the document]
Comment (corinnebosley) [reply .github/workflows/tests.yaml:72]: Cool.

**Pull request merged**